ambry-cloud/src/test/java/com.github.ambry.cloud/CloudBlobStoreTest.java

```diff
@@ -255,6 +278,156 @@ public void testExceptionalDest() throws Exception {
+    props.setProperty("clustermap.datacenter.name", "DC1");
+    props.setProperty("clustermap.ssl.enabled.datacenters", "DC0,DC1");
+    props.setProperty("clustermap.port", "12300");
+    props.setProperty("kms.default.container.key", TestUtils.getRandomKey(64));
+    props.setProperty("vcr.ssl.port", "12345");
+
+    ReplicationConfig replicationConfig = new ReplicationConfig(new VerifiableProperties(props));
+    ClusterMapConfig clusterMapConfig = new ClusterMapConfig(new VerifiableProperties(props));
+    CloudConfig cloudConfig = new CloudConfig(new VerifiableProperties(props));
+    CloudDataNode cloudDataNode = new CloudDataNode(cloudConfig, clusterMapConfig);
+
+    LatchBasedInMemoryCloudDestination latchBasedInMemoryCloudDestination =
+        new LatchBasedInMemoryCloudDestination(blobIdList);
+    CloudReplica cloudReplica = new CloudReplica(cloudConfig, partitionId, cloudDataNode);
+    CloudBlobStore cloudBlobStore =
+        new CloudBlobStore(new VerifiableProperties(props), partitionId, latchBasedInMemoryCloudDestination,
+            new VcrMetrics(new MetricRegistry()));
+    cloudBlobStore.start();
+
+    // Prepare RemoteReplicaInfo for ReplicaThread.
+    Map<DataNodeId, List<RemoteReplicaInfo>> replicasToReplicate = null;
+    for (ReplicaId replica : partitionId.getReplicaIds()) {
+      if (replica.getDataNodeId() == remoteHost.dataNodeId) {
+        RemoteReplicaInfo remoteReplicaInfo =
+            new RemoteReplicaInfo(replica, cloudReplica, cloudBlobStore, new MockFindToken(0, 0), Long.MAX_VALUE,
+                SystemTime.getInstance(), new Port(remoteHost.dataNodeId.getPort(), PortType.PLAINTEXT));
+        replicasToReplicate =
+            Collections.singletonMap(remoteHost.dataNodeId, Collections.singletonList(remoteReplicaInfo));
+        break;
+      }
+    }
+    ReplicationMetrics replicationMetrics = new ReplicationMetrics(new MetricRegistry(), Collections.emptyList());
+    replicationMetrics.populatePerColoMetrics(Collections.singleton(remoteHost.dataNodeId.getDatacenterName()));
+
+    ReplicaThread replicaThread =
+        new ReplicaThread("threadtest", replicasToReplicate, new MockFindToken.MockFindTokenFactory(), clusterMap,
+            new AtomicInteger(0), cloudDataNode, connectionPool, replicationConfig, replicationMetrics, null,
+            storeKeyConverter, transformer, clusterMap.getMetricRegistry(), false, cloudDataNode.getDatacenterName(),
+            new ResponseHandler(clusterMap), new MockTime());
+
+    long referenceTime = System.currentTimeMillis();
+    // Case 1: Put already expired. Replication happens after Put and after TtlUpdate.
+    // Upload to Cloud only after replicating ttlUpdate.
+    BlobId id = blobIdList.get(0);
+    addPutMessagesToReplicasOfPartition(id, accountId, containerId, partitionId, Collections.singletonList(remoteHost),
+        referenceTime - 2000, referenceTime - 1000);
+    replicaThread.replicate(new ArrayList<>(replicasToReplicate.values()));
```

Comment:
You can call replicasToReplicate.values() once and pass result the the refactored method.